src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                                var newMethodInvocation = generator.InvocationExpression(methodName, invocationArguments);
+                                var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+                                return editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+                            });
+                        }
+                    }
+
+                    if (document.Id == semanticDocument.Document.Id)
+                    {
+                        if (trampoline)
+                        {
+                            await IntroduceNewMethodAsync(semanticDocument, expression, methodSymbol, newMethodIdentifier, editor, cancellationToken).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            await IntroduceNewMethodOverloadAsync(semanticDocument, expression, methodSymbol, editor, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        UpdateExpressionInOriginalFunction(semanticDocument, expression, oldMethodDeclaration, parameterName, editor, allOccurrences, cancellationToken);
+
+                        var parameterType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type ?? semanticDocument.SemanticModel.Compilation.ObjectType;
+                        var refKind = syntaxFacts.GetRefKindOfArgument(expression);
+                        var parameter = generator.ParameterDeclaration(name: parameterName, type: generator.TypeExpression(parameterType), refKind: refKind);
+                        editor.InsertParameter(oldMethodDeclaration, insertionIndex, parameter);
+                    }
+
+                    var newRoot = editor.GetChangedRoot();
+                    modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(firstCallSite.Id, newRoot);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument
+        /// </summary>
+        private async Task<Solution> RewriteCallSitesAsync(SemanticDocument semanticDocument,
+                                                           TExpressionSyntax expression,
+                                                           ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites,
+                                                           bool allOccurrences,
+                                                           string parameterName,
+                                                           CancellationToken cancellationToken)
+        {
+            var expressionCopy = expression;
```

Comment:
needs docs to explain what you're doing.

Reply:
i still need this.  why are we tracking nodes at all?

Reply:
We are tracking the identifiers in the expression because when we go through and replace the node we need to get the current node based on the identifier. If it is not tracked then they throw an error that it doesn't exist,

Reply:
i don't undertand this.